accord-core/src/main/java/accord/impl/InMemoryCommandStore.java

```diff
@@ -837,44 +839,43 @@ public void shutdown() {}
 
     public static class SingleThread extends InMemoryCommandStore
     {
-        private final AtomicReference<Thread> expectedThread = new AtomicReference<>();
+        private Thread thread; // when run in the executor this will be non-null, null implies not running in this store
         private final ExecutorService executor;
 
-        public SingleThread(int id, NodeTimeService time, Agent agent, DataStore store, ProgressLog.Factory progressLogFactory, RangesForEpochHolder rangesForEpochHolder)
+        private SingleThread(int id, NodeTimeService time, Agent agent, DataStore store, ProgressLog.Factory progressLogFactory, RangesForEpochHolder rangesForEpochHolder)
         {
-            this(id, time, agent, store, progressLogFactory, rangesForEpochHolder, Executors.newSingleThreadExecutor(r -> {
+            super(id, time, agent, store, progressLogFactory, rangesForEpochHolder);
+            this.executor = Executors.newSingleThreadExecutor(r -> {
                 Thread thread = new Thread(r);
                 thread.setName(CommandStore.class.getSimpleName() + '[' + time.id() + ']');
                 return thread;
-            }));
-        }
-
-        private SingleThread(int id, NodeTimeService time, Agent agent, DataStore store, ProgressLog.Factory progressLogFactory, RangesForEpochHolder rangesForEpochHolder, ExecutorService executor)
-        {
-            super(id, time, agent, store, progressLogFactory, rangesForEpochHolder);
-            this.executor = executor;
+            });
+            executor.execute(() -> thread = Thread.currentThread());
         }
 
-        public static CommandStore.Factory factory(ExecutorService executor)
+        public static SingleThread create(int id, NodeTimeService time, Agent agent, DataStore store, ProgressLog.Factory progressLogFactory, RangesForEpochHolder rangesForEpochHolder)
```

Comment:
Minor refactor suggestion:

```
        interface Constructor<ST extends SingleThread>
        {
            ST construct(int id, NodeTimeService time, Agent agent, DataStore store, ProgressLog.Factory progressLogFactory, RangesForEpochHolder rangesForEpochHolder);
        }

        public static SingleThread create(int id, NodeTimeService time, Agent agent, DataStore store, ProgressLog.Factory progressLogFactory, RangesForEpochHolder rangesForEpochHolder)
        {
            return create(id, time, agent, store, progressLogFactory, rangesForEpochHolder, SingleThread::new);
        }

        protected static <ST extends SingleThread> ST create(int id, NodeTimeService time, Agent agent, DataStore store, ProgressLog.Factory progressLogFactory, RangesForEpochHolder rangesForEpochHolder, Constructor<ST> constructor)
        {
            ST st = constructor.construct(id, time, agent, store, progressLogFactory, rangesForEpochHolder);
            st.execute(() -> st.thread = Thread.currentThread());
            st.execute(() -> CommandStore.register(st));
            return st;
        }
```

and then use the same for `Debug` (this also requires making `thread` package-private, but it is probably overall cleaner this way)

Reply:
reading more on JMM it does seem that grouping `final` fields *before* this "leak" is safe from the final field point of view and may not be safe for non-final fields... in our case those are all local thread processing, so would be fine...

Given this, I can revert back to constructor and document this assumption.